File: myonlinefm/src/main/java/com/sec/myonlinefm/classificationprogram/mainfacefragment/OnLineFavFragment.java
package com.sec.myonlinefm.classificationprogram.mainfacefragment;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.sec.myonlinefm.R;
import com.sec.myonlinefm.classificationprogram.ChannelProgramActivity;
import com.sec.myonlinefm.classificationprogram.data.FavData;
import com.sec.myonlinefm.dbdata.MySQLHelper;

import java.io.BufferedInputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLConnection;
import java.util.ArrayList;
import java.util.List;

/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * to handle interaction events.
 * Use the {@link OnLineFavFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class OnLineFavFragment extends Fragment {
    // TODO: Rename parameter arguments, choose names that match
    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
    private static final String ARG_PARAM1 = "param1";
    private static final String ARG_PARAM2 = "param2";

    // TODO: Rename and change types of parameters
    private String mParam1;
    private String mParam2;

    public static final String TAG = "OnLineFavFragment";

    public OnLineFavFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param param1 Parameter 1.
     * @param param2 Parameter 2.
     * @return A new instance of fragment OnLineFavFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static OnLineFavFragment newInstance(String param1, String param2) {
        OnLineFavFragment fragment = new OnLineFavFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PARAM1, param1);
        args.putString(ARG_PARAM2, param2);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mParam1 = getArguments().getString(ARG_PARAM1);
            mParam2 = getArguments().getString(ARG_PARAM2);
        }
    }

    private View view;
    private ListView mFavListView;
    private List<FavData> mFavList;
    private SQLiteDatabase db;
    private MySQLHelper mySQLHelper;
    private FavListAdapter favListAdapter;
    private final int UPDATE_FAV_LIST_VIEW = 1;
    private View topBarActionBarView = null;
    private Context mContext;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        view = inflater.inflate(R.layout.fragment_on_line_fav, container, false);
        mFavListView = (ListView) view.findViewById(R.id.on_line_fav_list);
        mySQLHelper = MySQLHelper.getInstances();
        db = mySQLHelper.getWritableDatabase();
        mFavList = new ArrayList<>();
        Thread t = new Thread(GetFavData);
        t.start();
        return view;
    }

    @Override
    public void onDestroyView() {
        view = null;
        super.onDestroyView();
    }

    @Override
    public void onAttach(Context context) {
        mContext = context;
        super.onAttach(context);
    }

    @Override
    public void onDetach() {
        super.onDetach();
    }

    @SuppressLint("HandlerLeak")
    private Handler mHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
            if(msg.what == UPDATE_FAV_LIST_VIEW) {
                favListAdapter = new FavListAdapter();
                mFavListView.setAdapter(favListAdapter);
                mFavListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                    @Override
                    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                        Intent i = new Intent(mContext, ChannelProgramActivity.class);
                        i.putExtra("channel_id", mFavList.get(position).getChannelID());
                        i.putExtra("channel_name", mFavList.get(position).getChannelName());
                        startActivity(i);
                    }
                });
            }
        }
    };
    private Runnable GetFavData = new Runnable() {
        @Override
        public void run() {
            Cursor cursor = db.query("fav_channels",null, null, null, null, null, null);
            if (cursor.moveToFirst()) {
                do {
                    FavData favData = new FavData();
                    int cid = cursor.getInt(cursor.getColumnIndex("channel_id"));
                    int category_id = cursor.getInt(cursor.getColumnIndex("category_id"));
                    String name = cursor.getString(cursor.getColumnIndex("channel_name"));
                    String url = cursor.getString(cursor.getColumnIndex("channel_them_url"));
                    String podCasterName = cursor.getString(cursor.getColumnIndex("pod_caster_name"));
                    Log.d("bin1111.yang","favData : "+name+" "+cid+" "+url);
                    favData.setChannelID(cid);
                    favData.setCategoryID(category_id);
                    favData.setChannelName(name);
                    favData.setPodCasterName(podCasterName);
                    favData.setChannelIcon(getBitmap(url));
                    mFavList.add(favData);
                } while (cursor.moveToNext());
            }
            cursor.close();
            if(!mFavList.isEmpty())
                mHandler.sendEmptyMessage(UPDATE_FAV_LIST_VIEW);
        }
    };

    static class ViewHolder {
        ImageView mDemandChannelPic;

        TextView mDemandChannelTitle;

        TextView mPodCasterName;

        TextView mCategoryName;

        LinearLayout imgProgramIcon_layout;

        ImageView imgInfoIcon;
    }

    private class FavListAdapter extends BaseAdapter {

        @Override
        public int getCount() {
            return mFavList.size();
        }

        @Override
        public Object getItem(int position) {
            return mFavList.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(final int position, View convertView, ViewGroup parent) {
            final ViewHolder holder;

            if(convertView == null) {
                LayoutInflater inflater = (LayoutInflater) mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
                assert inflater != null;
                convertView = inflater.inflate(R.layout.fav_channel_item, null);
                holder = new ViewHolder();

                holder.mDemandChannelPic = convertView.findViewById(R.id.demand_channel_pic);
                holder.mDemandChannelTitle = convertView.findViewById(R.id.demand_channel_title);
                holder.mPodCasterName = convertView.findViewById(R.id.play_count);
                holder.imgProgramIcon_layout = convertView.findViewById(R.id.img_program_icon_layout);
                holder.mCategoryName = convertView.findViewById(R.id.context_total);
                holder.imgInfoIcon = convertView.findViewById(R.id.program_icon);
                convertView.setTag(holder);
            }  else {
                holder = (ViewHolder) convertView.getTag();//取出ViewHolder对象
            }

            holder.mDemandChannelPic.setImageBitmap(mFavList.get(position).getChannelIcon());
            holder.mDemandChannelTitle.setText(mFavList.get(position).getChannelName());
            holder.mPodCasterName.setText(mFavList.get(position).getPodCasterName());

            if(mFavList.get(position).getCategoryID() != 5)
                holder.mCategoryName.setText("点播节目");
            if(mFavList.get(position).getCategoryID() == 5)
                holder.mCategoryName.setText("直播节目");



            holder.imgInfoIcon.setImageResource(R.drawable.hybrid_radio_on_star);

            parent.setOnKeyListener(new View.OnKeyListener() {

                @Override
                public boolean onKey(View v, int keyCode, KeyEvent event) {
                    View selectedView = mFavListView.getSelectedView();

                    if (selectedView == null) {
                        return false;
                    }

                    holder.imgProgramIcon_layout = selectedView.findViewById(R.id.img_program_icon_layout);

                    if (keyCode == KeyEvent.KEYCODE_DPAD_RIGHT) {
                        if (event.getAction() == KeyEvent.ACTION_UP) {
                            holder.imgProgramIcon_layout.setPressed(true);
                        }
                        return true;
                    } else if ((keyCode == KeyEvent.KEYCODE_DPAD_CENTER) || (keyCode == KeyEvent.KEYCODE_ENTER)) {
                        if (holder.imgProgramIcon_layout.isPressed() && event.getAction() == KeyEvent.ACTION_UP) {
                            holder.imgProgramIcon_layout.performClick();
                        } else if (!holder.imgProgramIcon_layout.isPressed() && event.getAction() == KeyEvent.ACTION_UP) {
                            mFavListView.performItemClick(selectedView, mFavListView.getSelectedItemPosition(), mFavListView.getSelectedItemId());
                        }
                        return true;
                    } else if (keyCode == KeyEvent.KEYCODE_DPAD_LEFT) {
                        holder.imgProgramIcon_layout.setPressed(false);
                        mFavListView.setActivated(true);
                        return true;
                    } else {
                        holder.imgProgramIcon_layout.setPressed(false);
                        mFavListView.setActivated(true);
                        return false;
                    }
                }
            });

            holder.imgProgramIcon_layout.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    holder.imgInfoIcon.setImageResource(R.drawable.hybrid_radio_off_star);
                    db.delete("fav_channels", "channel_id = ?",
                            new String[]{String.valueOf(mFavList.get(position).getChannelID())});
                    Toast.makeText(mContext, "收藏移除", Toast.LENGTH_SHORT).show();
//                    mFavOnLineChannel.setImageResource(R.drawable.hybrid_radio_off_star);
                    mFavList.remove(mFavList.get(position));
                    notifyDataSetChanged();
                }
            });

            return convertView;
        }
    }
    private Bitmap getBitmap(String url) {
        if(url == null)
            return null;
        Bitmap bm = null;
        try {
            URL iconUrl = new URL(url);
            URLConnection conn = iconUrl.openConnection();
            HttpURLConnection http = (HttpURLConnection) conn;
            int length = http.getContentLength();
            conn.connect();
            // 获得图像的字符流
            InputStream is = conn.getInputStream();
            BufferedInputStream bis = new BufferedInputStream(is, length);
            bm = BitmapFactory.decodeStream(bis);
            bis.close();
            is.close();// 关闭流
            Log.d("bin1111.yang","getBitmap : "+url);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return bm;
    }

}

File: myonlinefm/src/main/java/com/sec/myonlinefm/classificationprogram/mainfacefragment/OnLineFavFragment.kt
package com.sec.myonlinefm.classificationprogram.mainfacefragment

import android.database.sqlite.SQLiteDatabase
import android.graphics.Bitmap
import com.sec.myonlinefm.R
import android.annotation.SuppressLint
import android.widget.AdapterView.OnItemClickListener

import com.sec.myonlinefm.classificationprogram.data.FavData
import android.graphics.BitmapFactory

import android.content.*

import android.os.*
import android.util.Log
import android.view.*
import android.widget.*
import androidx.fragment.app.Fragment
import com.sec.myonlinefm.classificationprogram.ChannelProgramActivity
import com.sec.myonlinefm.dbdata.MySQLHelper
import java.io.BufferedInputStream
import java.lang.Exception
import java.net.HttpURLConnection
import java.net.URL
import java.util.ArrayList

/**
 * A simple [Fragment] subclass.
 * Activities that contain this fragment must implement the
 * to handle interaction events.
 * Use the [OnLineFavFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class OnLineFavFragment : Fragment() {
    // TODO: Rename and change types of parameters
    private var mParam1: String? = null
    private var mParam2: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            mParam1 = arguments!!.getString(ARG_PARAM1)
            mParam2 = arguments!!.getString(ARG_PARAM2)
        }
    }

    internal var view: View? = null
    private var mFavListView: ListView? = null
    private var mFavList: MutableList<FavData?>? = null
    private var db: SQLiteDatabase? = null
    private var mySQLHelper: MySQLHelper? = null
    private var favListAdapter: FavListAdapter? = null
    private val UPDATE_FAV_LIST_VIEW = 1
    private val topBarActionBarView: View? = null
    private var mContext: Context? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        view = inflater.inflate(R.layout.fragment_on_line_fav, container, false)
        mFavListView = view!!.findViewById<View?>(R.id.on_line_fav_list) as ListView
        mySQLHelper = MySQLHelper.Companion.getInstances()
        db = mySQLHelper!!.getWritableDatabase()
        mFavList = ArrayList()
        val t = Thread(GetFavData)
        t.start()
        return view
    }

    override fun onDestroyView() {
        view = null
        super.onDestroyView()
    }

    override fun onAttach(context: Context?) {
        mContext = context
        super.onAttach(context)
    }

    override fun onDetach() {
        super.onDetach()
    }

    @SuppressLint("HandlerLeak")
    private val mHandler: Handler? = object : Handler() {
        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            if (msg.what == UPDATE_FAV_LIST_VIEW) {
                favListAdapter = FavListAdapter()
                mFavListView!!.setAdapter(favListAdapter)
                mFavListView!!.setOnItemClickListener(OnItemClickListener { parent, view, position, id ->
                    val i = Intent(mContext, ChannelProgramActivity::class.java)
                    i.putExtra("channel_id", mFavList!!.get(position)!!.getChannelID())
                    i.putExtra("channel_name", mFavList!!.get(position)!!.getChannelName())
                    startActivity(i)
                })
            }
        }
    }
    private val GetFavData: Runnable? = Runnable {
        val cursor = db!!.query("fav_channels", null, null, null, null, null, null)
        if (cursor.moveToFirst()) {
            do {
                val favData = FavData()
                val cid = cursor.getInt(cursor.getColumnIndex("channel_id"))
                val category_id = cursor.getInt(cursor.getColumnIndex("category_id"))
                val name = cursor.getString(cursor.getColumnIndex("channel_name"))
                val url = cursor.getString(cursor.getColumnIndex("channel_them_url"))
                val podCasterName = cursor.getString(cursor.getColumnIndex("pod_caster_name"))
                Log.d("bin1111.yang", "favData : $name $cid $url")
                favData.channelID = cid
                favData.categoryID = category_id
                favData.channelName = name
                favData.podCasterName = podCasterName
                favData.channelIcon = getBitmap(url)
                mFavList!!.add(favData)
            } while (cursor.moveToNext())
        }
        cursor.close()
        if (!mFavList!!.isEmpty())
            mHandler!!.sendEmptyMessage(UPDATE_FAV_LIST_VIEW)
    }

    internal class ViewHolder {
        var mDemandChannelPic: ImageView? = null
        var mDemandChannelTitle: TextView? = null
        var mPodCasterName: TextView? = null
        var mCategoryName: TextView? = null
        var imgProgramIcon_layout: LinearLayout? = null
        var imgInfoIcon: ImageView? = null
    }

    private inner class FavListAdapter : BaseAdapter() {
        override fun getCount(): Int {
            return mFavList!!.size
        }

        override fun getItem(position: Int): Any? {
            return mFavList!!.get(position)
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View? {
            var convertView = convertView
            val holder: ViewHolder
            if (convertView == null) {
                val inflater = (mContext!!.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater)!!
                convertView = inflater.inflate(R.layout.fav_channel_item, null)
                holder = ViewHolder()
                holder.mDemandChannelPic = convertView.findViewById(R.id.demand_channel_pic)
                holder.mDemandChannelTitle = convertView.findViewById(R.id.demand_channel_title)
                holder.mPodCasterName = convertView.findViewById(R.id.play_count)
                holder.imgProgramIcon_layout = convertView.findViewById(R.id.img_program_icon_layout)
                holder.mCategoryName = convertView.findViewById(R.id.context_total)
                holder.imgInfoIcon = convertView.findViewById(R.id.program_icon)
                convertView.tag = holder
            } else {
                holder = convertView.tag as ViewHolder //取出ViewHolder对象
            }
            holder.mDemandChannelPic!!.setImageBitmap(mFavList!!.get(position)!!.getChannelIcon())
            holder.mDemandChannelTitle!!.setText(mFavList!!.get(position)!!.getChannelName())
            holder.mPodCasterName!!.setText(mFavList!!.get(position)!!.getPodCasterName())
            if (mFavList!!.get(position)!!.getCategoryID() != 5)
                holder.mCategoryName!!.setText("点播节目")
            if (mFavList!!.get(position)!!.getCategoryID() == 5)
                holder.mCategoryName!!.setText("直播节目")
            holder.imgInfoIcon!!.setImageResource(R.drawable.hybrid_radio_on_star)
            parent!!.setOnKeyListener(View.OnKeyListener { v, keyCode, event ->
                val selectedView = mFavListView!!.getSelectedView() ?: return@OnKeyListener false
                holder.imgProgramIcon_layout = selectedView.findViewById(R.id.img_program_icon_layout)
                if (keyCode == KeyEvent.KEYCODE_DPAD_RIGHT) {
                    if (event.action == KeyEvent.ACTION_UP) {
                        holder.imgProgramIcon_layout!!.setPressed(true)
                    }
                    true
                } else if (keyCode == KeyEvent.KEYCODE_DPAD_CENTER || keyCode == KeyEvent.KEYCODE_ENTER) {
                    if (holder.imgProgramIcon_layout!!.isPressed() && event.action == KeyEvent.ACTION_UP) {
                        holder.imgProgramIcon_layout!!.performClick()
                    } else if (!holder.imgProgramIcon_layout!!.isPressed() && event.action == KeyEvent.ACTION_UP) {
                        mFavListView!!.performItemClick(selectedView,
                                mFavListView!!.getSelectedItemPosition(),
                                mFavListView!!.getSelectedItemId()
                        )
                    }
                    true
                } else if (keyCode == KeyEvent.KEYCODE_DPAD_LEFT) {
                    holder.imgProgramIcon_layout!!.setPressed(false)
                    mFavListView!!.setActivated(true)
                    true
                } else {
                    holder.imgProgramIcon_layout!!.setPressed(false)
                    mFavListView!!.setActivated(true)
                    false
                }
            })
            holder.imgProgramIcon_layout!!.setOnClickListener(View.OnClickListener {
                holder.imgInfoIcon!!.setImageResource(R.drawable.hybrid_radio_off_star)
                db!!.delete(
                        "fav_channels",
                        "channel_id = ?",
                        arrayOf<String?>(mFavList!!.get(position)!!.getChannelID().toString())
                )
                Toast.makeText(mContext, "收藏移除", Toast.LENGTH_SHORT).show()
                //                    mFavOnLineChannel.setImageResource(R.drawable.hybrid_radio_off_star);
                mFavList!!.remove(mFavList!!.get(position))
                notifyDataSetChanged()
            })
            return convertView
        }
    }

    private fun getBitmap(url: String?): Bitmap? {
        if (url == null) return null
        var bm: Bitmap? = null
        try {
            val iconUrl = URL(url)
            val conn = iconUrl.openConnection()
            val http = conn as HttpURLConnection
            val length = http.contentLength
            conn.connect()
            // 获得图像的字符流
            val `is` = conn.getInputStream()
            val bis = BufferedInputStream(`is`, length)
            bm = BitmapFactory.decodeStream(bis)
            bis.close()
            `is`.close() // 关闭流
            Log.d("bin1111.yang", "getBitmap : $url")
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return bm
    }

    companion object {
        // TODO: Rename parameter arguments, choose names that match
        // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
        private val ARG_PARAM1: String? = "param1"
        private val ARG_PARAM2: String? = "param2"
        val TAG: String? = "OnLineFavFragment"

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param param1 Parameter 1.
         * @param param2 Parameter 2.
         * @return A new instance of fragment OnLineFavFragment.
         */
        // TODO: Rename and change types and number of parameters
        fun newInstance(param1: String?, param2: String?): Fragment? {
            val fragment = OnLineFavFragment()
            val args = Bundle()
            args.putString(ARG_PARAM1, param1)
            args.putString(ARG_PARAM2, param2)
            fragment.arguments = args
            return fragment
        }
    }
}